File: src/net/dinkla/raytracer/math/Point3D.scala
package net.dinkla.raytracer.math

/**
  * Created by Dinkla on 10.07.2015.
  */
abstract class Point3D[@specialized(Float, Double) F](x: F, y: F, z: F)(implicit numType: Fractional[F]) extends Element3D[F](x, y, z)  {

   import numType.mkNumericOps         // -f

   def unary_-(): Point3D[F]           // unary minus

   def +(v: Vector3D[F]): Point3D[F]        = plus(v)     // Point plus Vector gives a Point
   def +(p: Point3D[F]): Vector3D[F]        = plus(p)     // Point plus Point gives a Vector
   def +(f: F): Point3D[F]                  = plus(f)

   def plus(v: Vector3D[F]): Point3D[F]     // Point plus Vector gives a Point
   def plus(p: Point3D[F]): Vector3D[F]     // Point plus Point gives a Vector
   def plus(f: F): Point3D[F]

   def -(v: Vector3D[F]): Point3D[F]        = plus(-v)    // Point minus Vector gives a Point
   def -(p: Point3D[F]): Vector3D[F]        = plus(-p)    // Point minus Point gives a Vector
   def -(f: F)                              = plus(-f)
   def minus(v: Vector3D[F]): Point3D[F]    = plus(-v)    // Point minus Vector gives a Point
   def minus(p: Point3D[F]): Vector3D[F]    = plus(-p)    // Point minus Point gives a Vector
   def minus(f: F)                          = plus(-f)

  def toVector: Vector3D[F]

  override def equals(obj: scala.Any): Boolean
    = obj match {
    case that: Point3D[F] => x == that.x && y == that.y && z == that.z
    case _ => false
  }

  override def hashCode(): Int = (x, y, z).hashCode()

}

object Point3D {

  def Point3D(x: Double, y: Double, z: Double) = new Point3DD(x, y, z)

  def Point3D(x: Float, y: Float, z: Float) = new Point3DF(x, y, z)

  val ORIGIN: Point3DF = new Point3DF(0, 0, 0)
  val MAX: Point3DF = new Point3DF(Float.PositiveInfinity, Float.PositiveInfinity, Float.PositiveInfinity)
  val MIN: Point3DF = new Point3DF(Float.NegativeInfinity, Float.NegativeInfinity, Float.NegativeInfinity)
  val DEFAULT_CAMERA: Point3DF = new Point3DF(0, 10, 10)

}

File: src/main/kotlin/net/dinkla/raytracer/math/Point3D.kt
package net.dinkla.raytracer.math

import kotlin.Double.Companion.NEGATIVE_INFINITY
import kotlin.Double.Companion.POSITIVE_INFINITY

class Point3D(x: Double, y: Double, z: Double) : Element3D(x, y, z) {

    constructor(e: Element3D) : this(e.x, e.y, e.z)

    operator fun unaryMinus() = Vector3D(-x, -y, -z)

    operator fun plus(v: Vector3D) = Point3D(x + v.x, y + v.y, z + v.z)

    operator fun plus(f: Double) = Point3D(x + f, y + f, z + f)

    operator fun minus(v: Point3D): Vector3D = Vector3D(x - v.x, y - v.y, z - v.z)

    operator fun minus(v: Vector3D) = Point3D(x - v.x, y - v.y, z - v.z)

    operator fun minus(f: Double) = Point3D(x - f, y - f, z - f)

    override fun equals(other: Any?): Boolean {
        val p = other as? Point3D ?: return false
        return super.equals(p)
    }

    companion object {
        val UNIT = Point3D(1.0, 1.0, 1.0)
        val ORIGIN = Point3D(0.0, 0.0, 0.0)
        val MAX = Point3D(POSITIVE_INFINITY, POSITIVE_INFINITY, POSITIVE_INFINITY)
        val MIN = Point3D(NEGATIVE_INFINITY, NEGATIVE_INFINITY, NEGATIVE_INFINITY)
        val X = Point3D(1.0, 0.0, 0.0)
        val Y = Point3D(0.0, 1.0, 0.0)
        val Z = Point3D(0.0, 0.0, 1.0)
    }

}

infix operator fun Double.times(p: Point3D) = Point3D(this * p.x, this * p.y, this * p.z)
